### PR TITLE
[WIP] Allow clientside validation of dialog fields [Finishes #146312755]

### PR DESCRIPTION
**Pull request by chalettu**
PV https://www.pivotaltracker.com/story/show/146312755

### CODE REVIEW

**Comment by chalettu** [comment]
@miq-bot add_label fine/no
@miq-bot add_label catalogs
@miq-bot add_label enhancement

**Comment by chalettu** [comment]
@miq-bot add_label ux/review

**Comment by chalettu** [comment]
Screen shots of what the clientside validation errors look like.  
regex_field

**Comment by chalettu** [comment]
This screenshot shows validation on submit of a form.  
addtocard

**Comment by Loicavenel** [comment]
This is very nice. the only missing information is why is not valid but this will come from the other requirements where we can put an information that will provide some information when clicking on it or mouse over it.

**Comment by chalettu** [comment]
@Loicavenel , I created two messages for errors , "field doesn't match required format" and "field is required".   This goes back to our discussion last week with the greater issue of how do we deal with translating error messages.  I chose the "field doesn't match required format" message as a default for regex errors since we don't currently support any type of custom error messages at the api level and I couldn't think of something to really clearly indicate, "User got the regex wrong".

**Comment by chriskacerguis** [comment]
@ManageIQ/core-ux for your review :)

/cc @Loicavenel

File: client/app/shared/dialog-content/dialog-content.component.js [hunk 1]
```diff
@@ -52,8 +56,11 @@ function DialogContentController(API_BASE, lodash) {
     if (Object.keys(vm.parsedOptions).length > 0) {
       item.dialog_fields.forEach(iterateDialogFields);
     }
+    item.dialog_fields.forEach(setFieldDefaults);
```
Comment (AllenBW) [inline client/app/shared/dialog-content/dialog-content.component.js]: 59 to 62 could be simplified with .map and =>

```
item.dialog_fields.map((item)=>{return item.fieldValidation =null});
```

File: client/app/states/catalogs/details/details.state.js [hunk 2]
```diff
@@ -108,7 +108,10 @@ function Controller(dialogs, serviceTemplate, EventNotifications, DialogFieldRef
     if (!ShoppingCart.allowed()) {
       return;
     }
-
+    const allFieldsValidate = DialogUtilities.validateAllFields(allDialogFields);
+    if (!allFieldsValidate) {
```
Comment (AllenBW) [inline client/app/states/catalogs/details/details.state.js]: consider combining the if from line 108 with this one...

```
if (!ShoppingCart.allowed() && !allFieldsValidate){
return;
}
```

File: client/app/shared/dialog-content/dialog-content.component.js (hunk 1 above)
Comment (AllenBW) [inline client/app/shared/dialog-content/dialog-content.component.js]: consider using .map and => for 59-62

```
item.dialog_fields.map((item)=>{item.fieldValidation = null;});
```

File: client/assets/sass/_forms.sass [hunk 3]
```diff
@@ -1,3 +1,4 @@
+@import 'colors'
```
Comment (AllenBW) [inline client/assets/sass/_forms.sass]: this should be 'app_colors' (which is the only file that imports colors directly)

File: client/app/core/dialog-utilities.service.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  };
+
+  return service;
+
+  function validateField(field) {
+    const fieldValue = field.default_value;
+    if (field.required) {
+      if (fieldValue === '') {
+        field.fieldValidation = false;
+        field.errorMessage = __('field is required');
```
Comment (AllenBW) [inline client/app/core/dialog-utilities.service.js]: Leading with lowercase doesn't conform to patternfly validation example presented [here](http://www.patternfly.org/pattern-library/forms-and-controls/errors-and-validation/)

File: client/app/core/dialog-utilities.service.spec.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
+describe('Dialog Utilities - ', () => {
```
Comment (AllenBW) [inline client/app/core/dialog-utilities.service.spec.js]: Test name might be incomplete?  Unsure as to the meaning of the - at the end...

**Comment by chriskacerguis** [comment]
@chalettu we need to add a better "description" so people know what is wrong

**Comment by chriskacerguis** [comment]
Closing this, as Chris has stated that this will be rolled into the fix with the dialog merge to the ui-components.

**Pull request closed without merging**